packages/lesswrong/server/useractivities/cron.ts

```diff
@@ -0,0 +1,293 @@
+      const updateQuery = `
+        WITH new_activity AS (
+          SELECT * FROM (VALUES ${tempTableValues}) AS t(user_id, activity_array)
+        )
+        UPDATE "UserActivities"
+        SET
+          "activityArray" = (array_cat(new_activity.activity_array, "UserActivities"."activityArray"))[:${ACTIVITY_WINDOW_HOURS}],
+          "startDate" = $1,
+          "endDate" = $2
+        FROM new_activity
+        WHERE
+          "UserActivities"."visitorId" = new_activity.user_id
+          AND "UserActivities"."type" = '${visitorIdType}';
+      `;
+      await dataDb.none(updateQuery, [paddedStartDate.toISOString(), updateEndDate.toISOString()]);
+    }
+  }
+
+  // Third case: Zero-pad the start of rows for users who were previously active but have no new activity
+  // (i.e. they have existing rows in the table but have not been active since we last updated the table)
+  const inactiveExistingUserIds = existingUserIds.filter(id => !existingUsersData.some(({ userOrClientId }) => userOrClientId === id));
+  if (inactiveExistingUserIds.length > 0) {
+    const newActivityHours = Math.ceil((updateEndDate.getTime() - updateStartDate.getTime()) / (1000 * 60 * 60));
+    const inactiveExistingUserIdsChunked = chunk(1000, inactiveExistingUserIds)
+    console.log(`Adding zero-padding to ${inactiveExistingUserIds.length} existing rows for which there is no new activity, in ${inactiveExistingUserIdsChunked.length} chunks`)
+
+    for (const userIdsChunk of inactiveExistingUserIdsChunked) {
+      const updateQuery = `
+        UPDATE "UserActivities"
+        SET
+          "activityArray" = (array_cat(ARRAY[${Array(newActivityHours).fill(0).join(', ')}]::double precision[], "UserActivities"."activityArray"))[:${ACTIVITY_WINDOW_HOURS}],
+          "startDate" = $1,
+          "endDate" = $2
+        WHERE
+          "UserActivities"."visitorId" IN (${userIdsChunk.map((_, index) => `$${index + 3}`).join(', ')})
+          AND "UserActivities"."type" = '${visitorIdType}';
+      `;
+      const queryParams = [paddedStartDate.toISOString(), updateEndDate.toISOString(), ...userIdsChunk];
+      await dataDb.none(updateQuery, queryParams);
+    }
+  }
+
+  // Fourth case: Every user has had their activity updated now. Remove rows for users who are no longer active in the last
+  // ACTIVITY_WINDOW_HOURS (i.e. the activity array is all zeros)
+  const countQuery = `
+    SELECT COUNT(*) as delete_count FROM "UserActivities"
+    WHERE
+      "UserActivities"."type" = '${visitorIdType}'
+      AND array_position("UserActivities"."activityArray", 1) IS NULL; -- no "1" in the array
+  `;
+  const deleteQuery = `
+    DELETE FROM "UserActivities"
+    WHERE
+      "UserActivities"."type" = '${visitorIdType}'
+      AND array_position("UserActivities"."activityArray", 1) IS NULL; -- no "1" in the array
+  `;
+  const count = (await dataDb.one(countQuery)).delete_count;
```

Comment:
[No need to action] Can you get rid of the separate count query through a postgres returning number of rows affected thing?